stabilize `int_error_matching`

ethanboxx opened the pull request:
- PR: https://github.com/rust-lang/rust/pull/55705
- Tracking issue: #22639

The original PR was my first `std` contribution. Since then a few people have called for it to be stabilized. See discussion on https://github.com/rust-lang/rust/issues/22639

A significant number of people seem to be using this feature in their projects https://github.com/search?p=2&q=%23%21%5Bfeature%28int_error_matching%29%5D&type=Code

# Unanswered questions

> If this is to be stabilized, I think that the terminology has to be improved: `IntErrorKind` uses `Underflow` to refer to negative overflow, but underflow usually means when the absolute value of a floating-point number is so small it becomes 0 (or a subnormal).

_Originally posted by @tspiteri in https://github.com/rust-lang/rust/issues/22639#issuecomment-520167905_

> Maybe `PosOverflow` and `NegOverflow`?

_Originally posted by @JarrettBillingsley in https://github.com/rust-lang/rust/issues/22639#issuecomment-539685390_

I personally don't think it needs to change as in this position underflow is clearly associated with integers and seems consistent with other uses of the word underflow in `std`. I am interested to hear others thoughts.

rust-highfive commented:
r? @KodrAus

(rust_highfive has picked a reviewer for you, use r? to override)

KodrAus commented:
I definitely don’t think we should call the variant `Underflow`, because even though we’re not really consistent internally, the term does refer to a specific failure mode that doesn’t apply here. I don’t think that’s a reason to repurpose the term, just so that we don’t leave any doubt for callers trying to interpret these error messages.

In general though I think this is a useful stabilization! I’ll kick off the FCP and add my concern for underflow to let others weigh in.

@rfcbot fcp merge

KodrAus commented:
@rfcbot concern underflow vs negative overflow

See https://github.com/rust-lang/rust/pull/76455#issuecomment-689388789 I think we should use an accurate name for the negative overflow variant, rather than repurposing the term _underflow_, which has a specific meaning for floating point numbers.

withoutboats commented:
I've always used the term "underflow" for ints wrapping beyond their lower bound and didn't realize this was not common usage.

KodrAus commented:
I actually didn’t even know _underflow_ had a special meaning and used it the same way until I think I got assigned a docs PR once to use _overflow_ for integers that would exceed their upper or lower bound. It turns out [_underflow_](https://en.m.wikipedia.org/wiki/Arithmetic_underflow) specifically refers to floating point numbers that are too close to zero to distinguish from it.

bors commented:
:umbrella: The latest upstream changes (presumably #76804) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Darksonn commented:
This was discussed on the users forum: [link](https://users.rust-lang.org/t/misleading-error-in-str-parse-for-int-types/49178).

ethanboxx commented:
I will open a new PR soon addressing these issues rather than continuing this one.

I will also need to take a good look at the parsing code to ensure it produces the correct errors.

As I am rather busy right now I will have an update within the next few weeks.

KodrAus commented:
Thanks for working through all this @ethanboxx!

If you don't need to make many changes then we could probably continue on this PR, since it's already got its FCP going.

ethanboxx commented:
Closing in favour of #77640
